File: agent/server/src/diff.ts
import type { LockfileObject } from '@pnpm/lockfile.types'
import { getFilePathByModeInCafs, type PackageFileInfo, type PackageFilesIndex } from '@pnpm/store.cafs'
import type { StoreIndex } from '@pnpm/store.index'

import type { MissingFile, ResponseMetadata } from './protocol.js'

export interface IntegrityEntry {
  decoded: PackageFilesIndex
  rawBuffer: Uint8Array
}

/**
 * Build an index mapping integrity hashes to their PackageFilesIndex
 * (decoded for diff computation) and raw msgpack buffer (for sending to client).
 */
export function buildIntegrityIndex (storeIndex: StoreIndex): Map<string, IntegrityEntry> {
  const index = new Map<string, IntegrityEntry>()
  for (const [key, value] of storeIndex.entries()) {
    const tabIdx = key.indexOf('\t')
    if (tabIdx === -1) continue
    const integrity = key.slice(0, tabIdx)
    if (index.has(integrity)) continue
    const rawBuffer = storeIndex.getRaw(key)
    if (!rawBuffer) continue
    index.set(integrity, {
      decoded: value as PackageFilesIndex,
      rawBuffer,
    })
  }
  return index
}

export interface DiffResult {
  metadata: ResponseMetadata
  missingFiles: MissingFile[]
  /** Pre-packed msgpack buffers for each package, keyed by depPath */
  packageIndexBuffers: Map<string, { integrity: string, rawBuffer: Uint8Array }>
}

/**
 * Given a resolved lockfile, the client's store integrities, and the server's
 * integrity index, compute which files need to be sent to the client.
 *
 * The algorithm:
 * 1. Union all file digests from the client's existing packages
 * 2. For each package in the lockfile, look up its file index
 * 3. Any file digest NOT in the client's set is "missing" and must be sent
 * 4. Dedup within the response (same digest only sent once)
 */
export function computeDiff (
  lockfile: LockfileObject,
  storeIntegrities: string[],
  integrityIndex: Map<string, IntegrityEntry>,
  storeDir: string
): DiffResult {
  // 1. Build the set of file digests the client already has
  const clientDigests = new Set<string>()
  const clientIntegrities = new Set(storeIntegrities)

  for (const integrity of storeIntegrities) {
    const entry = integrityIndex.get(integrity)
    if (!entry) continue
    for (const [, fileInfo] of getFilesEntries(entry.decoded)) {
      const executable = (fileInfo.mode & 0o111) !== 0
      clientDigests.add(`${fileInfo.digest}:${executable ? 'x' : ''}`)
    }
  }

  // 2. Iterate resolved packages and compute missing files
  const packageIndexBuffers = new Map<string, { integrity: string, rawBuffer: Uint8Array }>()
  const missingFiles: MissingFile[] = []

  let totalPackages = 0
  let alreadyInStore = 0
  let packagesToFetch = 0
  let filesInNewPackages = 0
  let filesAlreadyInCafs = 0
  let filesToDownload = 0
  let downloadBytes = 0

  for (const [depPath, pkgSnapshot] of Object.entries(lockfile.packages ?? {})) {
    totalPackages++
    const integrity = getSnapshotIntegrity(pkgSnapshot)
    if (!integrity) continue

    // Client already has this exact package — skip entirely
    if (clientIntegrities.has(integrity)) {
      alreadyInStore++
      continue
    }

    const entry = integrityIndex.get(integrity)
    if (!entry) continue // package not indexed on server yet

    packagesToFetch++

    for (const [, fileInfo] of getFilesEntries(entry.decoded)) {
      filesInNewPackages++

      // Dedup by digest + executable flag — the same content may need to be
      // stored at both the exec and non-exec CAFS paths if different packages
      // reference it with different file modes.
      const executable = (fileInfo.mode & 0o111) !== 0
      const dedupeKey = `${fileInfo.digest}:${executable ? 'x' : ''}`
      if (!clientDigests.has(dedupeKey)) {
        clientDigests.add(dedupeKey)
        filesToDownload++
        downloadBytes += fileInfo.size
        missingFiles.push({
          digest: fileInfo.digest,
          size: fileInfo.size,
          executable,
          cafsPath: getFilePathByModeInCafs(storeDir, fileInfo.digest, fileInfo.mode),
        })
      } else {
        filesAlreadyInCafs++
      }
    }

    packageIndexBuffers.set(depPath, { integrity, rawBuffer: entry.rawBuffer })
  }

  return {
    metadata: {
      lockfile,
      stats: {
        totalPackages,
        alreadyInStore,
        packagesToFetch,
        filesInNewPackages,
        filesAlreadyInCafs,
        filesToDownload,
        downloadBytes,
      },
    },
    missingFiles,
    packageIndexBuffers,
  }
}

export function getFilesEntries (pkgIndex: PackageFilesIndex): Array<[string, PackageFileInfo]> {
  const { files } = pkgIndex
  if (files instanceof Map) return [...files.entries()]
  return Object.entries(files as Record<string, PackageFileInfo>)
}

function getSnapshotIntegrity (pkgSnapshot: { resolution?: { integrity?: string } | string }): string | undefined {
  if (!pkgSnapshot.resolution) return undefined
  if (typeof pkgSnapshot.resolution === 'string') return undefined
  return pkgSnapshot.resolution.integrity
}

File: pnpr/crates/pnpr/src/agent/diff.rs
//! File-level diff between a resolved dependency tree and what the
//! client already has in its content-addressable store.
//!
//! Port of the pnpm-agent TypeScript `diff.ts`. Given a resolved
//! lockfile, the client's store integrities, and the server's store
//! index, it computes which individual files the client is missing —
//! deduplicated by `(digest, executable)` — plus the per-package
//! msgpack index entries the client needs to write into its own store
//! index before a headless install.

use std::collections::{HashMap, HashSet};

use pacquet_store_dir::{
    PackageFilesIndex, StoreIndex, StoreIndexError, encode_package_files_index,
};

/// One resolved package the diff considers, distilled from the
/// lockfile's `packages` map.
pub struct ResolvedPackage {
    /// SRI integrity string, e.g. `sha512-...`.
    pub integrity: String,
    /// Package id without the peer-deps suffix, e.g. `foo@1.0.0`.
    pub pkg_id: String,
}

/// A file the client's store is missing.
pub struct MissingFile {
    /// Lowercase sha512 hex digest (no `sha512-` prefix).
    pub digest: String,
    pub size: u64,
    pub executable: bool,
}

/// Pre-packed store-index entry forwarded to the client (`I` line).
pub struct PackageIndexEntry {
    pub integrity: String,
    pub pkg_id: String,
    /// msgpackr-records bytes ready for the client's `StoreIndex.setRawMany`.
    pub raw: Vec<u8>,
}

#[derive(Default)]
pub struct Stats {
    pub total_packages: u64,
    pub already_in_store: u64,
    pub packages_to_fetch: u64,
    pub files_in_new_packages: u64,
    pub files_already_in_cafs: u64,
    pub files_to_download: u64,
    pub download_bytes: u64,
}

pub struct DiffResult {
    pub missing_files: Vec<MissingFile>,
    pub package_index: Vec<PackageIndexEntry>,
    pub stats: Stats,
}

struct IntegrityEntry {
    decoded: PackageFilesIndex,
    /// Re-encoded msgpackr-records buffer for the client.
    raw: Vec<u8>,
}

fn is_executable(mode: u32) -> bool {
    mode & 0o111 != 0
}

/// Build a map from SRI integrity to its decoded files index and a
/// re-encoded msgpackr-records buffer, restricted to the integrities
/// the diff actually needs (the client's existing packages plus the
/// newly resolved ones). Re-encoding guarantees the buffer is in the
/// msgpackr-records shape the pnpm client's store index reads, no
/// matter whether pacquet wrote the row as plain msgpack.
///
/// The server's store index keys are `{integrity}\t{pkgId}`; we key by
/// the integrity half and keep the first occurrence, matching
/// `buildIntegrityIndex` in the TypeScript agent.
fn build_integrity_index(
    store: &StoreIndex,
    needed: &HashSet<String>,
) -> Result<HashMap<String, IntegrityEntry>, StoreIndexError> {
    let mut index = HashMap::new();
    for key in store.keys()? {
        let Some((integrity, _pkg_id)) = key.split_once('\t') else { continue };
        if !needed.contains(integrity) || index.contains_key(integrity) {
            continue;
        }
        let Some(decoded) = store.get(&key)? else { continue };
        let Ok(raw) = encode_package_files_index(&decoded) else { continue };
        index.insert(integrity.to_string(), IntegrityEntry { decoded, raw });
    }
    Ok(index)
}

/// Compute the file-level diff. Mirrors `computeDiff` in the
/// TypeScript agent: union the client's existing file digests, then
/// for every resolved package not already in the client's store emit
/// the files it doesn't yet have (deduped across the whole response).
pub fn compute_diff(
    store: &StoreIndex,
    packages: &[ResolvedPackage],
    store_integrities: &[String],
) -> Result<DiffResult, StoreIndexError> {
    let mut needed: HashSet<String> = store_integrities.iter().cloned().collect();
    for pkg in packages {
        needed.insert(pkg.integrity.clone());
    }
    let index = build_integrity_index(store, &needed)?;

    let client_integrities: HashSet<&str> = store_integrities.iter().map(String::as_str).collect();

    // Digests (and their exec flag) the client already has on disk.
    let mut client_digests: HashSet<(String, bool)> = HashSet::new();
    for integrity in store_integrities {
        let Some(entry) = index.get(integrity) else { continue };
        for file in entry.decoded.files.values() {
            client_digests.insert((file.digest.clone(), is_executable(file.mode)));
        }
    }

    let mut stats = Stats::default();
    let mut missing_files = Vec::new();
    let mut package_index = Vec::new();

    for pkg in packages {
        stats.total_packages += 1;

        if client_integrities.contains(pkg.integrity.as_str()) {
            stats.already_in_store += 1;
            continue;
        }

        let Some(entry) = index.get(&pkg.integrity) else { continue };
        stats.packages_to_fetch += 1;

        for file in entry.decoded.files.values() {
            stats.files_in_new_packages += 1;
            let executable = is_executable(file.mode);
            let key = (file.digest.clone(), executable);
            if client_digests.insert(key) {
                stats.files_to_download += 1;
                stats.download_bytes += file.size;
                missing_files.push(MissingFile {
                    digest: file.digest.clone(),
                    size: file.size,
                    executable,
                });
            } else {
                stats.files_already_in_cafs += 1;
            }
        }

        package_index.push(PackageIndexEntry {
            integrity: pkg.integrity.clone(),
            pkg_id: pkg.pkg_id.clone(),
            raw: entry.raw.clone(),
        });
    }

    Ok(DiffResult { missing_files, package_index, stats })
}
